### PR TITLE
Update the Android documentation for Godot 4.2

### PR DESCRIPTION
**Pull request by m4gr3d**
Draft for updates to the Android documentation for Godot 4.2:

- Updates the Android plugins documentation based on the refactor done for Godot 4.2 in https://github.com/godotengine/godot/pull/80740, https://github.com/godotengine/godot/pull/76821, https://github.com/godotengine/godot/pull/78958 and https://github.com/godotengine/godot/pull/81368
  - Supersedes https://github.com/godotengine/godot-docs/pull/7416
  - Addresses https://github.com/godotengine/godot/pull/78958#issuecomment-1684312579 and https://github.com/godotengine/godot/pull/78958#issuecomment-1656804456 and https://github.com/godotengine/godot/pull/80740#issuecomment-1683371175
- Adds a page to describes the Godot Android library and provide instructions for how to use it
  - Fixes https://github.com/godotengine/godot-docs/issues/4018 
- Fix misc issues, updates terminologies as needed

### Remaining TODOs (to be addressed in follow-up PRs):
- [ ] Incorporate the NDK location or rather the instructions to compile godot-cpp via scons for Android when the NDK is not at the root directory. See https://github.com/godotengine/godot-docs/pull/7884#issuecomment-1710298393
- [ ] Update the OpenXR documentation to provide information on how to build an Android plugin with OpenXR GDExtension capabilities

*Docs maintainers note: ping Yuri once merged*

### CODE REVIEW

File: tutorials/platform/android/android_library.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+by tapping into the functionality provided by Android platforms and ecosystem.
+
+They are themselves Android libraries with a dependency on the Godot Android library,
+which they use to access the Godot API and integrate into the engine's lifecycle.
+
+Their nature as Android libraries means they offer some of the same key features as the Godot Android library,
+notably the ability to integrate with the Gradle buildsystem, and the ability for the plugins themselves to be portable and embeddable.
+
+Their access to the Godot engine APIs and lifecycle grant them powerful capabilities
+such as GDExtension which allow them to update / mods the engine behavior as needed.
```
Comment (Anutrix) [inline tutorials/platform/android/android_library.rst:38]:
```suggestion
such as GDExtension which allow them to update / mod the engine behavior as needed.
```

File: tutorials/platform/android/android_in_app_purchases.rst [hunk 2]
```diff
@@ -1,17 +1,11 @@
-:article_outdated: True
-
 .. _doc_android_in_app_purchases:
 
 Android in-app purchases
 ========================
 
 Godot offers a first-party ``GodotGooglePlayBilling`` Android plugin compatible with Godot 3.2.2 and higher.
```
Comment (dsnopek) [inline tutorials/platform/android/android_in_app_purchases.rst:6]: Given that these are the 4.x docs, does it still make sense to reference "Godot 3.2.2 and higher" here?

File: tutorials/platform/android/android_plugin.rst [hunk 3]
```diff
@@ -1,165 +1,388 @@
...
+XR platforms without encumbering the core codebase with vendor specific code or binaries.
 
 Android plugin
 --------------
 
-While introduced in Godot 3.2, the Android plugin system got a significant architecture update starting with Godot 3.2.2.
-The new plugin system is backward-incompatible with the previous one and in Godot 4.0, the previous system was fully deprecated and removed.
-Since we previously did not version the Android plugin systems, the new one is now labelled ``v1`` and is the starting point for the modern Godot Android ecosystem.
+**Version 1 (v1)** of the Android plugin system was introduced in Godot 3.2.2. That version allowed developers
+to augment the Godot 3 engine with Java / Kotlin logic as well as GDNative functionality.
```
Comment (dsnopek) [inline tutorials/platform/android/android_plugin.rst:19]: Again, does it make sense to reference "Godot 3.2.2" and "GDNative" in the context of the Godot 4.x docs? Perhaps this should just talk about v1 being used in Godot 4.0 and 4.1?

File: tutorials/platform/android/android_plugin.rst [hunk 4]
```diff
@@ -1,165 +1,388 @@
...
+3. After updating the Godot Android library dependency, sync or build the plugin and resolve any compile errors:
+
+    - The ``Godot`` instance provided by ``GodotPlugin::getGodot()`` no longer has access to a ``android.content.Context`` reference. Use ``GodotPlugin::getActivity()`` instead.
+
+4. Delete the ``gdap`` configuration file(s) and follow the instructions in the `Packaging a v2 Android plugin`_ section to set up the plugin configuration.
+
+Packaging a v2 Android plugin
+-----------------------------
+
+As mentioned, a v2 Android plugin is now provided to the Godot Editor as an EditorExportPlugin plugin, so it shares a lot of the `same packaging steps <https://docs.godotengine.org/en/stable/tutorials/plugins/editor/making_plugins.html#creating-a-plugin>`_.
```
Comment (dsnopek) [inline tutorials/platform/android/android_plugin.rst:146]:
```suggestion
As mentioned, a v2 Android plugin is now provided to the Godot Editor as an ``EditorExportPlugin plugin``, so it shares a lot of the `same packaging steps <https://docs.godotengine.org/en/stable/tutorials/plugins/editor/making_plugins.html#creating-a-plugin>`_.
```

File: tutorials/platform/android/android_plugin.rst [hunk 5]
```diff
@@ -1,165 +1,388 @@
...
+            # Specifies which platform is supported by the plugin
+            func _supports_platform(platform):
+                if platform is EditorExportPlatformAndroid:
+                    return true
+                return false
+
+            # Return the paths relative to the 'addons' directory of the plugin's AAR binaries
+            func _get_android_libraries(platform, debug):
+                if debug:
+                    return <paths_to_debug_android_plugin_aar_binaries>
```
Comment (dsnopek) [inline tutorials/platform/android/android_plugin.rst:172]: From the sample project, these seem to need to be `PackedStringArray`s, so it might make sense to show this like:

```suggestion
                    return PackedStringArray(["<paths_to_debug_android_plugin_aar_binaries>"])
```

File: tutorials/platform/android/android_plugin.rst [hunk 6]
```diff
@@ -1,165 +1,388 @@
...
+                else:
+                    return <paths_to_release_android_plugin_aar_binaries>
```
Comment (dsnopek) [inline tutorials/platform/android/android_plugin.rst:174]: Same here:

```suggestion
                    return PackedStringArray(["<paths_to_release_android_plugin_aar_binaries>"])
```

**Review by dsnopek — COMMENTED** [review]
This is looking great to me! Just a couple notes from skimming the diff

File: tutorials/platform/android/android_plugin.rst (hunk 3 above)
Comment (m4gr3d) [reply tutorials/platform/android/android_plugin.rst:19]: This section is to give some background as to the reason for the `v2` versioning.
But to your point, I rephrase it to mention compatibility with Godot 4.0 and 4.1 and remove reference to GDNative.

File: tutorials/platform/android/android_plugin.rst [hunk 7]
```diff
@@ -1,165 +1,388 @@
...
+
+        - This is used to configure the Android plugin and hook it within the Godot Editor's export process. It should look something like this::
+
+            @tool
+            extends EditorExportPlugin
+
+            # Plugin's name
+            var _plugin_name = "<plugin_name>"
+
+            # Specifies which platform is supported by the plugin
```
Comment (AThousandShips) [inline tutorials/platform/android/android_plugin.rst:163]:
```suggestion
            # Specifies which platform is supported by the plugin.
```

File: tutorials/platform/android/android_plugin.rst [hunk 8]
```diff
@@ -1,165 +1,388 @@
...
+            func _supports_platform(platform):
+                if platform is EditorExportPlatformAndroid:
+                    return true
+                return false
+
+            # Return the paths relative to the 'addons' directory of the plugin's AAR binaries
```
Comment (AThousandShips) [inline tutorials/platform/android/android_plugin.rst:169]:
```suggestion
            # Return the paths relative to the 'addons' directory of the plugin's AAR binaries.
```

File: tutorials/platform/android/android_plugin.rst [hunk 9]
```diff
@@ -1,165 +1,388 @@
...
+            func _get_android_libraries(platform, debug):
+                if debug:
+                    return PackedStringArray(["<paths_to_debug_android_plugin_aar_binaries>"])
+                else:
+                    return PackedStringArray(["<paths_to_release_android_plugin_aar_binaries>"])
+
+            # Return the plugin's name
```
Comment (AThousandShips) [inline tutorials/platform/android/android_plugin.rst:176]:
```suggestion
            # Return the plugin's name.
```

File: tutorials/platform/android/android_plugin.rst [hunk 10]
```diff
@@ -1,165 +1,388 @@
...
+    |
+    +--plugin.gdextension           # GDExtension config file
+
+
+Here is what the ``plugin.gdextension`` config file should look like::
+
+    [configuration]
+
+    entry_symbol = "plugin_library_init"
+    compatibility_minimum = 4.2
```
Comment (AThousandShips) [inline tutorials/platform/android/android_plugin.rst:271]:
```suggestion
    compatibility_minimum = "4.2"
```
See https://github.com/godotengine/godot-docs/issues/7864

File: tutorials/platform/android/android_plugin.rst [hunk 11]
```diff
@@ -1,165 +1,388 @@
...
+provide one (or multiple) gdscript wrapper class for your plugin users to interface with.
+
+If using the plugin templates, those wrapper classes should be included in the `plugin/export_scripts_template/interface`
+directory (create the directory if it doesn't exist).
+
+For example::
+
+    class_name PluginInterface extends Object
+
+    ## Interface used to access the functionality provided by this plugin
```
Comment (AThousandShips) [inline tutorials/platform/android/android_plugin.rst:357]:
```suggestion
    ## Interface used to access the functionality provided by this plugin.
```

File: tutorials/platform/android/android_plugin.rst (hunk 10 above)
Comment (m4gr3d) [reply tutorials/platform/android/android_plugin.rst:271]: Thanks for the info! I need to update the project templates as well.

**Comment by paddy-exe** [comment]
Reminder from RocketChat to incorporate the NDK location or rather the instructions to compile godot-cpp via scons for Android when the NDK is not at the root directory. In this case the `ANDROID_NDK_ROOT` path has to be set. Mine was at `C:\Users\Patrick\AppData\Local\Android\Sdk\ndk\23.2.8568313`

File: tutorials/platform/android/android_plugin.rst (hunk 3 above)
Comment (dsnopek) [reply tutorials/platform/android/android_plugin.rst]: Looks good!

**Comment by tokengamedev** [comment]
I peeked into the documentation in the commit code. I am not sure I can read the documentation without this getting merged, but I tried. I have few comments, correct me if I am wrong somewhere, 

**Building and Packaging the Plugin**

1. We are dropping the local godot library aar file dependency in favor of maven dependency. I think it is better compared to V1, but can you confirm? Also, can you provide a beta version somewhere when engine beta version is released, as 4.2.0 stable is not available yet.

_On a side note, version 4.1.1 version is missing in the maven repo_

2. When we package the plugin, the output will be a plugin directory as follows (non-gdextension)
``` gdscript
+ MyPLugin
|
+--.bin/  # output location for the AAR binaries
|   |
|   +--MyPlugin.1.1.2.release.aar  # the release version of the aar file (naming standards is up to the creator)
|   |
|   +--MyPlugin.1.1.2.debug.aar  # the debug version of the aar file (naming standards is up to the creator)
|
+--.export/  # scripts directory
|   |
|   +--editor_export_plugin.gd  # EditorExportPlugin tool script
|   |
|   +--editor_plugin.gd # EditorPlugin tool script
|
+--plugin.cfg  # plugin INI file
```
Too much work for the plugin to be created, compared to just generating a gdap file, but I hope there are merits to it.

3. Can you provide simple examples or usage for all methods inside editor_export_plugin (like _get_android_dependencies_maven_repos and others)

 
**Using the plugin**

1. It seems like the plugin has to be enabled two times to be usable, one in project settings and one in export preset. Is that so?
2. When the plugin is imported, which name will be used to fetch the plugin Singleton in code
    Name in the plugin.cfg or 
    _get_name() function of the editor_export_plugin.gd

Thanks for the hard work of refactoring the plugin for android, I haven't started testing yet, let me know if there is any way I can get hold of godot library aar file so that I can start testing with simple plugins. 

👍 😎

**Comment by Rindbee** [comment]
I think it is necessary to modify `cmdline-tools;latest` (It is now `cmdline-tools;11.0` and does not support JDK11) to `cmdline-tools;10.0` for JDK11. See #7902.

**Comment by tokengamedev** [comment]
Is this going to be updated sometime soon? 
Anyway I am busy next couple of weeks working for the critical part of my next game.

BTW, I saw aar files in the Godot builds, I hope I can use them for dependency. Let me know.

**Comment by m4gr3d** [comment]
> Is this going to be updated sometime soon? Anyway I am busy next couple of weeks working for the critical part of my next game.
> 
> BTW, I saw aar files in the Godot builds, I hope I can use them for dependency. Let me know.

@tokengamedev yes, I'm working on incorporating the received feedback; expect an update this week.

**Comment by mhilbrunner** [comment]
@m4gr3d With Godot 4.2 beta imminent, what is the status of this? :)

**Comment by m4gr3d** [comment]
> @m4gr3d With Godot 4.2 beta imminent, what is the status of this? :)

@mhilbrunner, working on that today! 

I was blocked working on the sample apps to go alongside the documentation; I completed those yesterday (see https://github.com/godotengine/godot/pull/82893 and https://github.com/m4gr3d/Godot-Android-Samples/commit/94a648089c10a25c0761f915661c300206abe421) so I should be now unblocked.

**Comment by mhilbrunner** [comment]
Great to hear :) No pressure or anything, just wanted to check. Feel free to ping me once ready to review/merge!

**Comment by m4gr3d** [comment]
@tokengamedev I've addressed your questions / feedback below.

> I peeked into the documentation in the commit code. I am not sure I can read the documentation without this getting merged, but I tried. I have few comments, correct me if I am wrong somewhere,
> 

You can take a look as my branch as github should generate and make the documentation readable: https://github.com/m4gr3d/godot-docs/blob/update_android_plugin_tutorial/tutorials/platform/android/android_plugin.rst

> **Building and Packaging the Plugin**
> 
> 1. We are dropping the local godot library aar file dependency in favor of maven dependency. I think it is better compared to V1, but can you confirm? Also, can you provide a beta version somewhere when engine beta version is released, as 4.2.0 stable is not available yet.
> 

That's correct, maven dependency is the preferred and recommended approach, but the local godot library aar will continue to be provided for the time being for those that prefer that approach.

Until 4.2 is stable, you can use the snapshot versions available at https://s01.oss.sonatype.org/content/repositories/snapshots/org/godotengine/godot/4.2.0.dev-SNAPSHOT/. Those versions are updated for each dev / beta release.

To use the snapshot versions, you'll need to include the maven url for the snapshot repo. See [here for reference](https://github.com/m4gr3d/Godot-Android-Samples/blob/cb57f3f29f36bfaafa9f26140105dc16cd51a5a7/settings.gradle#L16).

> _On a side note, version 4.1.1 version is missing in the maven repo_
> 

This  has been fixed, and it should now be available.

> 2. When we package the plugin, the output will be a plugin directory as follows (non-gdextension)
> 
> ```gdscript
> + MyPLugin
> |
> +--.bin/  # output location for the AAR binaries
> |   |
> |   +--MyPlugin.1.1.2.release.aar  # the release version of the aar file (naming standards is up to the creator)
> |   |
> |   +--MyPlugin.1.1.2.debug.aar  # the debug version of the aar file (naming standards is up to the creator)
> |
> +--.export/  # scripts directory
> |   |
> |   +--editor_export_plugin.gd  # EditorExportPlugin tool script
> |   |
> |   +--editor_plugin.gd # EditorPlugin tool script
> |
> +--plugin.cfg  # plugin INI file
> ```
> 
> Too much work for the plugin to be created, compared to just generating a gdap file, but I hope there are merits to it.
> 

This has been reduced to the following:

```
    export_scripts_template/
    |
    +--export_plugin.gd         # export plugin tool script
    |
    +--plugin.cfg               # plugin INI file
```

So using your example, that would correspond to:

```
    MyPlugin/
    |
    +--MyPlugin.1.1.2.release.aar  # the release version of the aar file (naming standards is up to the creator)
    |
    +--MyPlugin.1.1.2.debug.aar  # the debug version of the aar file (naming standards is up to the creator)
    |
    +--export_plugin.gd         # export plugin tool script
    |
    +--plugin.cfg               # plugin INI file
```

So compared to the previous (.gdap) approach, only one additional script file is added.

> but I hope there are merits to it.

The merits are most evident for plugins that provide customization options. The previous (.gdap) approach had no such capabilities. With the new approach, v2 Android plugins can now:
- Include options in the Android export presets using [_get_export_options](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html#class-editorexportplugin-method-get-export-options), along with [warnings](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html#class-editorexportplugin-method-get-export-option-warning) to ensure the selected options match the plugin's preconditions
- Customize the `AndroidManifest` file for the generated binary in a manner that's preserved on editor upgrade using [_get_android_manifest_activity_element_contents](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html#class-editorexportplugin-method-get-android-manifest-activity-element-contents), [_get_android_manifest_application_element_contents](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html#class-editorexportplugin-method-get-android-manifest-application-element-contents) and [_get_android_manifest_element_contents](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html#class-editorexportplugin-method-get-android-manifest-element-contents)
- Customize the export flow using the rest of the APIs provided by the [EditorExportPlugin](https://docs.godotengine.org/en/latest/classes/class_editorexportplugin.html)

The Godot OpenXR Loaders plugin makes heavy use of the new capabilities to provide options, capabilities and plugin binaries for multiple OpenXR vendors. Take a look at [its implementation for reference](https://github.com/GodotVR/godot_openxr_loaders/tree/master/demo/addons/godotopenxr).

> 3. Can you provide simple examples or usage for all methods inside editor_export_plugin (like _get_android_dependencies_maven_repos and others)
> 

The Godot OpenXR Loaders plugin is a good reference for usage of the methods:
- https://github.com/m4gr3d/godot_openxr_loaders/blob/update_plugin_directory_structure/demo/addons/godotopenxr/export/godot_openxr_export_plugin.gd
- https://github.com/m4gr3d/godot_openxr_loaders/blob/update_plugin_directory_structure/demo/addons/godotopenxr/export/meta/godot_openxr_meta_editor_export_plugin.gd

> **Using the plugin**
> 
> 1. It seems like the plugin has to be enabled two times to be usable, one in project settings and one in export preset. Is that so?

No, only the first time in the project settings is required. Once the plugin is enabled in the project settings, it's then up to the plugin developer whether additional checks are required or not. 
For example, the Godot OpenXR Loaders plugin supports multiple OpenXR vendors, so I include additional options in the export preset to let the user select which vendors should be enabled. The plugin also includes warnings to prevent users from enabling more than 1 vendor for a single export preset.

> 2. When the plugin is imported, which name will be used to fetch the plugin Singleton in code
>    Name in the plugin.cfg or
>    _get_name() function of the editor_export_plugin.gd
> 

The name used is the one returned by [GodotPlugin#getPluginName()](https://github.com/godotengine/godot/blob/6916349697a4339216469e9bf5899b983d78db07/platform/android/java/lib/src/org/godotengine/godot/plugin/GodotPlugin.java#L276). 
I'd recommend the value in `plugin.cfg` and `_get_name()` function to also match that value (the provided github template do so), but it's not required.

> Thanks for the hard work of refactoring the plugin for android, I haven't started testing yet, let me know if there is any way I can get hold of godot library aar file so that I can start testing with simple plugins.
> 
> 👍 😎

Feel free to take a look at https://github.com/m4gr3d/Godot-Android-Samples/tree/master, a collection of samples I develop for reference.

File: _tools/redirects/redirects.csv [hunk 12]
```diff
@@ -130,7 +130,7 @@ source,destination
 /getting_started/workflow/best_practices/scene_organization.html,/tutorials/best_practices/scene_organization.html
 /getting_started/workflow/best_practices/scenes_versus_scripts.html,/tutorials/best_practices/scenes_versus_scripts.html
 /getting_started/workflow/best_practices/what_are_godot_classes.html,/tutorials/best_practices/what_are_godot_classes.html
-/getting_started/workflow/export/android_custom_build.html,/tutorials/export/android_custom_build.html
+/getting_started/workflow/export/android_custom_build.html,/tutorials/export/android_gradle_build.html
```
Comment (mhilbrunner) [inline _tools/redirects/redirects.csv:133]: This means redirects would break for 4.0 and 4.1. I *think* adding a new second redirect with this instead should work fine, hopefully. Not 100% sure, it's been some time since messing with the redirects.

File: tutorials/platform/android/android_library.rst [hunk 13]
```diff
@@ -0,0 +1,240 @@
+.. _doc_android_library:
+
+Godot Android Library
+=====================
+
+The Godot Engine for Android platforms is designed to be used as an `Android library <https://developer.android.com/studio/projects/android-library>`_.
+This architecture enables several key features on Android platforms:
+
+- Ability to integrate the Gradle buildsystem within the Godot Editor which provides the ability to leverage more components from the Android ecosystem such as libraries and tools
```
Comment (mhilbrunner) [inline tutorials/platform/android/android_library.rst:9]:
```suggestion
- Ability to integrate the Gradle build system within the Godot Editor, which provides the ability to leverage more components from the Android ecosystem such as libraries and tools
```

**Review by mhilbrunner — CHANGES_REQUESTED** [review]
The PNGs need to be converted to WebP for newly added/changed images (see [image creation guidelines](https://docs.godotengine.org/en/stable/contributing/documentation/docs_image_guidelines.html)).

Besides that and the other small things I noted above, I spotted nothing. Great work :)

File: _tools/redirects/redirects.csv (hunk 12 above)
Comment (m4gr3d) [reply _tools/redirects/redirects.csv:133]: > Alternatively (and potentially simpler) we could rename the page in 4.0 and 4.1 too and then have this redirect everywhere in 4.x, if it applies to all versions?

This change would apply to all 4.x versions so I can split this specific change in its own PR and add a cherry-pick for 4.0 and 4.1 as well.

**Comment by m4gr3d** [comment]
@mhilbrunner I've addressed your feedback!

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Merged 🎉 Great work!

File: _tools/redirects/redirects.csv (hunk 12 above)
Comment (m4gr3d) [reply _tools/redirects/redirects.csv:133]: @mhilbrunner I've backported the changes in https://github.com/godotengine/godot-docs/pull/8217

**Comment by tokengamedev** [comment]
@m4gr3d thanks for the update, sorry could not reply earlier as been busy with the game code 🤓. I will try to check out this week. 
Appreciate your great work ❤️
